accounting/invoice/utils.py

```diff
@@ -21,27 +21,115 @@
 
-def get_first_day_past_month():
+    Example without using a new dictionary:
+    >>> d = {
+    >>>     'a': lambda a: a**2
+    >>>     'b': 'str'
+    >>> }
+    >>> resolve_dict_callables(d, 4)
+    >>> print(d)
+    >>> # We lose the callable.
+    >>> {'a': 16, 'b': 'str'}
     """
-    Return the `datetime` object representing the first day of the month that just passed.
-    """
-    return get_last_day_past_month().replace(day=1)
+    new_dict = {}
+    for key, potential_callable in dictionary.items():
+        if callable(potential_callable):
+            new_dict[key] = potential_callable(*args)
+        else:
+            new_dict[key] = dictionary[key]
+    return new_dict
 
 
-def default_invoice_number():
+# pylint: disable=too-many-arguments,too-many-locals
+def send_email_with_invoice(template, subject, extra_email_context=None, client_usernames=settings.BILLING_CYCLE_USERS,
+                            create_invoice=False, draft_invoice=False, fill_line_items_from_jira=False,
+                            upload_to_google_drive=False, auto_approve=False):
     """
-    Return the invoice number which defaults to `yyyy-mm` for the current year and month.
+    Send an email with an invoice attached, and potentially perform peripheral actions.
+
+    TODO: This function's pretty beefy; break it up by encapsulating logic into more utility functions or a new class.
+
+    :param template: The email template.
+    :param subject: The email subject.
+    :param extra_email_context: Extra kwargs to update the email context with.
+    :param client_usernames: The list of billing cycle users' usernames.
+    :param create_invoice: Whether to create a new invoice based off of a past one.
+    :param draft_invoice: Whether this is a draft invoice or not.
+    :param fill_line_items_from_jira: Whether to fill up the invoice's line items from JIRA worklogs.
+    :param upload_to_google_drive: Whether to upload the resulting invoice PDF to Google Drive.
+                                   If using a draft invoice, it gets uploaded into `invoices-in/draft/{month}`.
     """
-    return timezone.now().strftime("%Y-%m")
+    if extra_email_context is None:
+        extra_email_context = {}
+    now = datetime.datetime.now()
+    past = get_last_day_past_month()
+    past_month_formatted = past.strftime('%B')
+    template = get_template(template)
+    for client_username in client_usernames:
+        client = Account.objects.get(user__username=client_username)
+        providers = [rate.provider for rate in client.client_hourly_rates.all()]
+        for provider in providers:
+            # Copy over certain details from the latest invoice from the proper month to the new invoice.
+            invoice = models.Invoice.objects.filter(
+                provider=provider,
+                client=client,
+                date__month=past.month if draft_invoice else now.month,
+            ).latest('date')
```

Comment:
@UmanShahzad When running `send_invoice_approval_request` this fails with `DoesNotExist: Invoice matching query does not exist.` if there is no invoice for the previous month, which is an edge case that we'll need to cover. (Everybody's getting paid for the first time at some point :))

Reply:
@itsjeyd @antoviaque Yep, I ran into this already, but the issue is that we need some kind of 'template' of the invoice because it has to copy over `extra_text`, `extra_image`, etc., which would *have* to be copied from somewhere.

This would require a significant refactor, to use a new `InvoiceTemplate` class for each provider that contains all of these fields, which would have a 1-to-many relationship with `Invoice`s themselves, so that they can pull information from the template. This way for new team members we'd just need to create the template and then the invoice will use that for its per-provider info.

Since that's going to be a lot of changes (mostly names and custom migration code) and I only thought about it today, then for all new team members who use our automation, Alicia will have to create an initial dummy paid invoice with the template details. This dummy invoice, if created for the previous month and already marked paid, won't have an effect on our transferwise bulk payments, and won't have any line items associated with them. When we move to using real templates we can delete the dummy invoices.

I can work on this next week, but it's the last week before the billing cycle, so no guarantee that it'll be deployed to prod by the 1st.

Reply:
@UmanShahzad Rather than a whole separate model, these could be fields on the provider object for now? Ie `provider.default invoice_extra_text`, etc. - this way at least the current PR would have a proper behavior, without weirdness like these that we'll need to remember, and without adding too much refactoring work. Then the InvoiceTemplate could be implemented in the future if/when we need to add more similar fields or custom templating behavior.

Reply:
@antoviaque Hmm.. instead, I'll just make `InvoiceTemplate`, but skip any migration work. It's equivalent to what you're suggesting, except it builds the initial code infrastructure for the model, which is a good step forward, and that invoices would not be linked to it: each `Account` (i.e. provider) will be linked to it. Then in the future we can remove that relation in place of 1-to-many with InvoiceTemplate-to-Invoice.

Reply:
I went the mile: implementing the whole thing right now properly.